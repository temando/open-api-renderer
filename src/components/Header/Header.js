import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import Description from '../Description/Description'
import ExternalLink from '../ExternalLink/ExternalLink'
import './Header.scss'

export default class Header extends PureComponent {
  render () {
    const { title, version, description, info, specUrl } = this.props

    return (
      <header id='header'>
        <h1>{title}</h1>

        <dl className='inline-pairs'>
          <dt>Version</dt>
          <dd>{version}</dd>
          {info && info.contact && this.renderContact(info.contact)}
          {info && info.license && this.renderLicense(info.license)}
        </dl>

        {description && <Description description={description} />}

        <nav>
          <ExternalLink href={specUrl}>Download OpenAPI definition</ExternalLink>
          {info && info.termsOfService &&
          <ExternalLink href={info.termsOfService}>Terms of Service</ExternalLink>}
        </nav>
      </header>
    )
  }

  /**
   * The contact object has no required fields, so the logic is a bit meh as
   * they are several permutations to consider.
   *
   * @param {Object} contact
   */
  renderContact (contact) {
    const elements = [<dt key='contact'>Contact</dt>]

    if (contact.name && contact.url) {
      elements.push(
        <dd key='contact-value'>
          <ExternalLink href={contact.url}>{contact.name}</ExternalLink>
          {contact.email && `(${contact.email})`}
        </dd>
      )
    } else if (contact.name) {
      elements.push(
        <dd key='contact-value'>{contact.name}</dd>
      )
    } else if (contact.url) {
      elements.push(
        <dd key='contact-value'>
          <ExternalLink href={contact.url}>{contact.url}</ExternalLink>
          {contact.email && `(${contact.email})`}
        </dd>
      )
    } else if (contact.email) {
      elements.push(
        <dd key='contact-value'>
          <ExternalLink href={`mailto:${contact.email}`}>{contact.email}</ExternalLink>
        </dd>
      )
    }

    return elements
  }

  renderLicense (license) {
    return [
      <dt key='license'>License</dt>,
      <dd key='license-value'>
        {license.url && <ExternalLink href={license.url}>{license.name}</ExternalLink>}
        {!license.url && license.name}
      </dd>
    ]
  }
}

Header.propTypes = {
  title: PropTypes.string.isRequired,
  version: PropTypes.string.isRequired,
  description: PropTypes.string,
  info: PropTypes.shape({
    contact: PropTypes.object,
    license: PropTypes.object,
    termsOfService: PropTypes.string
  }),
  specUrl: PropTypes.string.isRequired
}
