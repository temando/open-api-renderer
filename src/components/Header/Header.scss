@import '../../base.scss';

#header {
  padding: 0 20px;

  h1 {
    margin-bottom: .5rem;
  }

  nav a {
    display: inline-block;
    margin-right: .5rem;
  }
}
