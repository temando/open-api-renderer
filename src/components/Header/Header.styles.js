export const styles = theme => ({
  'header': {
    padding: '0 2rem',

    '& h1': {
      margin: '.5rem 0'
    },

    '& nav a': {
      display: 'inline-block',
      marginRight: '.5rem'
    }
  },

  'inlinePairs': {
    display: 'inline-flex',
    margin: '1em 0',
    padding: '0',
    borderBottom: `1px solid ${theme.borders.default}`,

    '& dt, & dd': {
      margin: '0',
      padding: '.5rem',
      borderTop: `1px solid ${theme.borders.default}`
    },

    '& dt': {
      paddingRight: '1.5rem',
      fontWeight: 600
    }
  },

  [`@media (max-width: ${theme.sizes.breakpoint})`]: {
    'header': {
      padding: '0 1rem'
    },
    'inlinePairs': {
      display: 'block'
    }
  }
})
