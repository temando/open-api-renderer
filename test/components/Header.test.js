import React from 'react'
import Header from './../../src/components/Header/Header'
import renderer from 'react-test-renderer'

describe('<Header />', () => {
  const title = 'Swagger Petstore'
  const version = '1.0.0'
  const specUrl = 'https://temando.github.io/open-api-renderer/petstore-open-api-v3.0.0-RC1.json'

  it('renders header with minimum props', () => {
    const tree = renderer.create(
      <Header title={title} version={version} specUrl={specUrl} />
    )

    expect(tree).toMatchSnapshot()
  })

  it('renders license information (name)', () => {
    const info = {
      license: {
        name: 'Apache 2.0'
      }
    }
    const tree = renderer.create(
      <Header title={title} version={version} specUrl={specUrl} info={info} />
    )

    expect(tree).toMatchSnapshot()
  })

  it('renders license information (name, url)', () => {
    const info = {
      license: {
        name: 'Apache 2.0',
        url: 'http://www.apache.org/licenses/LICENSE-2.0.html'
      }
    }
    const tree = renderer.create(
      <Header title={title} version={version} specUrl={specUrl} info={info} />
    )

    expect(tree).toMatchSnapshot()
  })

  it('renders contact information (name)', () => {
    const info = {
      contact: {
        name: 'ACME Company'
      }
    }
    const tree = renderer.create(
      <Header title={title} version={version} specUrl={specUrl} info={info} />
    )

    expect(tree).toMatchSnapshot()
  })

  it('renders contact information (url)', () => {
    const info = {
      contact: {
        url: 'http://example.org'
      }
    }
    const tree = renderer.create(
      <Header title={title} version={version} specUrl={specUrl} info={info} />
    )

    expect(tree).toMatchSnapshot()
  })

  it('renders contact information (email)', () => {
    const info = {
      contact: {
        email: 'hello@example.org'
      }
    }
    const tree = renderer.create(
      <Header title={title} version={version} specUrl={specUrl} info={info} />
    )

    expect(tree).toMatchSnapshot()
  })

  it('renders contact information (name, url)', () => {
    const info = {
      contact: {
        name: 'ACME Company',
        url: 'http://example.org'
      }
    }
    const tree = renderer.create(
      <Header title={title} version={version} specUrl={specUrl} info={info} />
    )

    expect(tree).toMatchSnapshot()
  })

  it('renders contact information (name, url, email)', () => {
    const info = {
      contact: {
        name: 'ACME Company',
        url: 'http://example.org',
        email: 'hello@example.org'
      }
    }
    const tree = renderer.create(
      <Header title={title} version={version} specUrl={specUrl} info={info} />
    )

    expect(tree).toMatchSnapshot()
  })
})
