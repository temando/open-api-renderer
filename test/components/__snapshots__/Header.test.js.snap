// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> renders contact information (email) 1`] = `
<header
  className={undefined}
  id="header"
>
  <h1>
    Swagger Petstore
  </h1>
  <dl
    className={undefined}
  >
    <dt>
      Version
    </dt>
    <dd>
      1.0.0
    </dd>
    <dt>
      Contact
    </dt>
    <dd>
      <a
        href="mailto:hello@example.org"
        rel="noopener noreferrer"
        target="_blank"
      >
        hello@example.org
      </a>
    </dd>
  </dl>
  <nav>
    <a
      href="https://temando.github.io/open-api-renderer/petstore-open-api-v3.0.0-RC1.json"
      rel="noopener noreferrer"
      target="_blank"
    >
      Download OpenAPI definition
    </a>
  </nav>
</header>
`;

exports[`<Header /> renders contact information (name) 1`] = `
<header
  className={undefined}
  id="header"
>
  <h1>
    Swagger Petstore
  </h1>
  <dl
    className={undefined}
  >
    <dt>
      Version
    </dt>
    <dd>
      1.0.0
    </dd>
    <dt>
      Contact
    </dt>
    <dd>
      ACME Company
    </dd>
  </dl>
  <nav>
    <a
      href="https://temando.github.io/open-api-renderer/petstore-open-api-v3.0.0-RC1.json"
      rel="noopener noreferrer"
      target="_blank"
    >
      Download OpenAPI definition
    </a>
  </nav>
</header>
`;

exports[`<Header /> renders contact information (name, url) 1`] = `
<header
  className={undefined}
  id="header"
>
  <h1>
    Swagger Petstore
  </h1>
  <dl
    className={undefined}
  >
    <dt>
      Version
    </dt>
    <dd>
      1.0.0
    </dd>
    <dt>
      Contact
    </dt>
    <dd>
      <a
        href="http://example.org"
        rel="noopener noreferrer"
        target="_blank"
      >
        ACME Company
      </a>
    </dd>
  </dl>
  <nav>
    <a
      href="https://temando.github.io/open-api-renderer/petstore-open-api-v3.0.0-RC1.json"
      rel="noopener noreferrer"
      target="_blank"
    >
      Download OpenAPI definition
    </a>
  </nav>
</header>
`;

exports[`<Header /> renders contact information (name, url, email) 1`] = `
<header
  className={undefined}
  id="header"
>
  <h1>
    Swagger Petstore
  </h1>
  <dl
    className={undefined}
  >
    <dt>
      Version
    </dt>
    <dd>
      1.0.0
    </dd>
    <dt>
      Contact
    </dt>
    <dd>
      <a
        href="http://example.org"
        rel="noopener noreferrer"
        target="_blank"
      >
        ACME Company
      </a>
      (hello@example.org)
    </dd>
  </dl>
  <nav>
    <a
      href="https://temando.github.io/open-api-renderer/petstore-open-api-v3.0.0-RC1.json"
      rel="noopener noreferrer"
      target="_blank"
    >
      Download OpenAPI definition
    </a>
  </nav>
</header>
`;

exports[`<Header /> renders contact information (url) 1`] = `
<header
  className={undefined}
  id="header"
>
  <h1>
    Swagger Petstore
  </h1>
  <dl
    className={undefined}
  >
    <dt>
      Version
    </dt>
    <dd>
      1.0.0
    </dd>
    <dt>
      Contact
    </dt>
    <dd>
      <a
        href="http://example.org"
        rel="noopener noreferrer"
        target="_blank"
      >
        http://example.org
      </a>
    </dd>
  </dl>
  <nav>
    <a
      href="https://temando.github.io/open-api-renderer/petstore-open-api-v3.0.0-RC1.json"
      rel="noopener noreferrer"
      target="_blank"
    >
      Download OpenAPI definition
    </a>
  </nav>
</header>
`;

exports[`<Header /> renders description 1`] = `
<header
  className={undefined}
  id="header"
>
  <h1>
    Swagger Petstore
  </h1>
  <dl
    className={undefined}
  >
    <dt>
      Version
    </dt>
    <dd>
      1.0.0
    </dd>
  </dl>
  <div
    className=""
    dangerouslySetInnerHTML={
      Object {
        "__html": "<p>Hello world</p>
",
      }
    }
  />
  <nav>
    <a
      href="https://temando.github.io/open-api-renderer/petstore-open-api-v3.0.0-RC1.json"
      rel="noopener noreferrer"
      target="_blank"
    >
      Download OpenAPI definition
    </a>
  </nav>
</header>
`;

exports[`<Header /> renders docs 1`] = `
<header
  className={undefined}
  id="header"
>
  <h1>
    Swagger Petstore
  </h1>
  <dl
    className={undefined}
  >
    <dt>
      Version
    </dt>
    <dd>
      1.0.0
    </dd>
  </dl>
  <nav>
    <a
      href="https://temando.github.io/open-api-renderer/petstore-open-api-v3.0.0-RC1.json"
      rel="noopener noreferrer"
      target="_blank"
    >
      Download OpenAPI definition
    </a>
  </nav>
  <div
    className=""
  >
    <div
      className=""
      dangerouslySetInnerHTML={
        Object {
          "__html": "<p>Hello</p>
",
        }
      }
    />
    <a
      href="http://example.org"
      rel="noopener noreferrer"
      target="_blank"
    >
      More information Â»
    </a>
  </div>
</header>
`;

exports[`<Header /> renders header with minimum props 1`] = `
<header
  className={undefined}
  id="header"
>
  <h1>
    Swagger Petstore
  </h1>
  <dl
    className={undefined}
  >
    <dt>
      Version
    </dt>
    <dd>
      1.0.0
    </dd>
  </dl>
  <nav>
    <a
      href="https://temando.github.io/open-api-renderer/petstore-open-api-v3.0.0-RC1.json"
      rel="noopener noreferrer"
      target="_blank"
    >
      Download OpenAPI definition
    </a>
  </nav>
</header>
`;

exports[`<Header /> renders license information (name) 1`] = `
<header
  className={undefined}
  id="header"
>
  <h1>
    Swagger Petstore
  </h1>
  <dl
    className={undefined}
  >
    <dt>
      Version
    </dt>
    <dd>
      1.0.0
    </dd>
    <dt>
      License
    </dt>
    <dd>
      Apache 2.0
    </dd>
  </dl>
  <nav>
    <a
      href="https://temando.github.io/open-api-renderer/petstore-open-api-v3.0.0-RC1.json"
      rel="noopener noreferrer"
      target="_blank"
    >
      Download OpenAPI definition
    </a>
  </nav>
</header>
`;

exports[`<Header /> renders license information (name, url) 1`] = `
<header
  className={undefined}
  id="header"
>
  <h1>
    Swagger Petstore
  </h1>
  <dl
    className={undefined}
  >
    <dt>
      Version
    </dt>
    <dd>
      1.0.0
    </dd>
    <dt>
      License
    </dt>
    <dd>
      <a
        href="http://www.apache.org/licenses/LICENSE-2.0.html"
        rel="noopener noreferrer"
        target="_blank"
      >
        Apache 2.0
      </a>
    </dd>
  </dl>
  <nav>
    <a
      href="https://temando.github.io/open-api-renderer/petstore-open-api-v3.0.0-RC1.json"
      rel="noopener noreferrer"
      target="_blank"
    >
      Download OpenAPI definition
    </a>
  </nav>
</header>
`;
